client/lib/browser-storage/index.ts

```diff
@@ -0,0 +1,137 @@
+/**
+ * Internal dependencies
+ */
+import {
+	getStoredItem as bypassGet,
+	setStoredItem as bypassSet,
+	clearStorage as bypassClear,
+} from './bypass';
+
+let shouldBypass = false;
+
+const DB_NAME = 'calypso';
+const DB_VERSION = 2; // Match versioning of the previous localforage-based implementation.
+const STORE_NAME = 'calypso_store';
+
+const supportsIDB = typeof window !== 'undefined' && window.indexedDB;
+
+let cachedDbPromise: Promise< IDBDatabase > | null = null;
+
+function prepareIDB() {
+	const request = indexedDB.open( DB_NAME, DB_VERSION );
+	return new Promise< IDBDatabase >( ( resolve, reject ) => {
+		if ( request ) {
+			request.onerror = () => reject( request.error );
+			request.onsuccess = () => resolve( request.result );
+			request.onupgradeneeded = () => request.result.createObjectStore( STORE_NAME );
+		}
+	} );
+}
+
+function getDB() {
+	cachedDbPromise = cachedDbPromise || prepareIDB();
+	return cachedDbPromise;
+}
+
+/**
+ * Whether persistent storage should be bypassed, using a memory store instead.
+ *
+ * @param shouldBypassPersistentStorage Whether persistent storage should be bypassed.
+ */
+export function bypassPersistentStorage( shouldBypassPersistentStorage: boolean ) {
+	shouldBypass = shouldBypassPersistentStorage;
+}
+
+/**
+ * Get a stored item.
+ *
+ * @param key The stored item key.
+ * @return A promise with the stored value. `null` if missing.
+ */
+export function getStoredItem< T >( key: string ): Promise< T | null > {
+	if ( shouldBypass ) {
+		return bypassGet( key );
+	}
+
+	if ( ! supportsIDB ) {
+		const valueString = localStorage.getItem( key );
+		return Promise.resolve( valueString ? JSON.parse( valueString ) : null );
```

Comment:
`JSON.parse` can fail and should result in `getStoredItem` returning a rejected Promise rather than throwing.

Reply:
Good point. This is now handled.